fix-point/app/Main.hs

```diff
@@ -1,77 +1,60 @@
 
+import Control.Applicative (optional)
 import Data.Semigroup ((<>))
-import Options.Applicative (help, Parser, ParserInfo, info, (<**>), header,
-                            helper, execParser, (<|>), metavar, long, short,
-                            fullDesc, progDesc, argument, str)
-import System.Directory (doesDirectoryExist)
+import Options.Applicative (Parser, long, metavar, help, helper, progDesc,
+                            fullDesc, header, info, (<**>), execParser, strOption)
+import System.Directory (doesDirectoryExist, getCurrentDirectory)
 
 import CopyFiles (copyAll)
 import Rename
 
 main :: IO ()
-main = run =<< execParser (args `withInfo` "Create project with given name")
-
-run :: Args -> IO ()
-run (Args input) = do
-    case input of
-        Project  proj           -> copyAll "three-layer" proj proj
-        ProjPref proj pref      -> copyAll "three-layer" proj pref
-        StringFP path           -> copyAll path "NEWNAME" "NEWDIR"
-        Combo    path proj pref -> copyAll path proj pref
-
-type Project    = String
-type Prefix     = String
-type SourcePath = FilePath
-
-data FPInput = Project Project
-             | ProjPref Project Prefix
-             | StringFP SourcePath
-             | Combo SourcePath Project Prefix
-
-data Args = Args FPInput
-
-args :: Parser Args
-args = Args <$> parseFPInput
-
-parseFPInput :: Parser FPInput
-parseFPInput = project <|> projPref <|> stringFP <|> combo
-
-project :: Parser FPInput
-project = Project
-    <$> argument str (metavar "PROJECT NAME" <> help "name of project" )
-
-projPref :: Parser FPInput
-projPref = ProjPref
-    <$> argument str
-        (metavar "PROJECT NAME" <> help "name of project" )
-    <*> argument str
-        (  long "prefix"
-        <> short 'p'
-        <> metavar "PREFIX NAME"
-        <> help "name of prefix" )
-
-stringFP :: Parser FPInput
-stringFP = StringFP <$> argument str
-           (  long "source path"
-           <> short 's'
-           <> metavar "SOURCEPATH"
-           <> help "source path" )
-
-combo :: Parser FPInput
-combo = Combo
-    <$> argument str
-    (  long "source path"  -- This line gave a 'no instance error'... 
-    <> short 's'
+main = test =<< execParser opts
+  where
+    opts = info (parseOptions <**> helper)
+      (  fullDesc
+      <> progDesc "Test parsing of CLI's"
+      <> header "Test for optparse-applicative")
+
+test :: Options -> IO ()
+test opts = case opts of
+    (Options proj (Just pref) (Just sd)) -> copyAll sd proj pref
+    (Options proj Nothing (Just sd))     -> copyAll sd proj proj
+    (Options proj (Just pref) Nothing)   -> copyAll "three-layer" proj pref
+    (Options proj Nothing Nothing)       -> copyAll "three-layer" proj proj
+test _ = return ()
+
+data Options = Options
+    { projectName :: String
+    , prefixName :: Maybe String
+    , sourceDirectory :: Maybe String
+    }
+
+parseOptions :: Parser Options
+parseOptions = Options <$> parseProject <*> parsePref <*> parseSD
```

Comment:
You don't need to create separate parsers for `Maybe`. Just write:

```haskell
parseOptions = Options <$> parseProject <*> optional parsePref <*> optional parseSD
```

You can even it align it to make it look cooler:

```haskell
parseOptions = Options
    <$> parseProject
    <*> optional parsePref
    <*> optional parseSD
```